Scripts/Runtime/Data/VirtualData/VirtualDataLookup.cs

```diff
@@ -0,0 +1,112 @@
+            where TInstance : unmanaged, IKeyedData<TKey>
+        {
+            Type type = typeof(VirtualData<TKey, TInstance>);
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+            if (m_DataLookup.ContainsKey(type))
+            {
+                throw new InvalidOperationException($"VirtualData of type {type} has already been added!");
+            }
+#endif
+            m_DataLookup.Add(type, data);
+            m_Data.Add(data);
+        }
+        
+        /// <summary>
+        /// Returns <see cref="VirtualData{TKey,TInstance}"/> from the lookup
+        /// </summary>
+        /// <typeparam name="TKey">The type of Key</typeparam>
+        /// <typeparam name="TInstance">The type of Instance data</typeparam>
+        /// <returns>The <see cref="VirtualData{TKey,TInstance}"/> instance</returns>
+        public VirtualData<TKey, TInstance> GetData<TKey, TInstance>()
+            where TKey : unmanaged, IEquatable<TKey>
+            where TInstance : unmanaged, IKeyedData<TKey>
+        {
+            Type type = typeof(VirtualData<TKey, TInstance>);
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+            if (!m_DataLookup.ContainsKey(type))
+            {
+                throw new InvalidOperationException($"VirtualData of type {type} has not been added! Please ensure {nameof(AddData)} has been called.");
+            }
+#endif
+            
+            return (VirtualData<TKey, TInstance>)m_DataLookup[type];
+        }
+        
+        /// <summary>
+        /// Consolidates all <see cref="VirtualData{TKey,TInstance}"/> in the lookup in parallel.
+        /// </summary>
+        /// <param name="dependsOn">The <see cref="JobHandle"/> consolidation work depends on.</param>
+        /// <returns>
+        /// A <see cref="JobHandle"/> that represents when all <see cref="VirtualData{TKey,TInstance}"/>
+        /// consolidation is complete.
+        /// </returns>
+        public JobHandle ConsolidateForFrame(JobHandle dependsOn)
+        {
+            return m_VirtualDataBulkScheduler.BulkSchedule(dependsOn);
+        }
+        
+        //*************************************************************************************************************
+        // BULK SCHEDULERS
+        //*************************************************************************************************************
+        
+        /// <summary>
+        /// <see cref="AbstractBulkScheduler{T}"/> to call ConsolidateForFrame on <see cref="VirtualData{TKey,TInstance}"/>
+        /// </summary>
+        private class VirtualDataBulkSchedulerForConsolidate : AbstractBulkScheduler<AbstractVirtualData>
+        {
+            public VirtualDataBulkSchedulerForConsolidate(List<AbstractVirtualData> list) : base(list)
+            {
+            }
+
+            protected override JobHandle ScheduleItem(AbstractVirtualData item, JobHandle dependsOn)
+            {
+                return item.ConsolidateForFrame(dependsOn);
+            }
+        }
```

Comment:
Hmmm, I know passing a delegate like this will cause an allocation
`SomeMethod(SomeDelegate)`

But you could cache the delegate in a field
```
private readonly Action m_SomeDelegate = SomeDelegate;
// ...
SomeMethod(m_SomeDelegate)
```

This avoids the allocation. Up to you whether that's actually cleaner.

Reply:
Probably no reflection needed. This has kind of merged with this comment. https://github.com/decline-cookies/anvil-unity-dots/pull/40#discussion_r913241088

Reply:
Re-implemented this. See `BulkSchedulingExtension`, `BulkSchedulingUtil`, `BulkScheduleDelegate`.